source/extensions/filters/http/jwt_authn/jwks_async_fetcher.cc

```diff
@@ -21,15 +27,25 @@ JwksAsyncFetcher::JwksAsyncFetcher(const RemoteJwks& remote_jwks,
                                    CreateJwksFetcherCb create_fetcher_fn,
                                    JwtAuthnFilterStats& stats, JwksDoneFetched done_fn)
     : remote_jwks_(remote_jwks), context_(context), create_fetcher_fn_(create_fetcher_fn),
-      stats_(stats), done_fn_(done_fn), cache_duration_(getCacheDuration(remote_jwks)),
+      stats_(stats), done_fn_(done_fn),
       debug_name_(absl::StrCat("Jwks async fetching url=", remote_jwks_.http_uri().uri())) {
   // if async_fetch is not enabled, do nothing.
   if (!remote_jwks_.has_async_fetch()) {
     return;
   }
 
-  cache_duration_timer_ =
-      context_.mainThreadDispatcher().createTimer([this]() -> void { fetch(); });
+  // Next refetch time should be: jwks_cache_duration - RefetchBeforeExpiredSec.
+  // Using RefetchBeforeExpiredSec is to avoid on_demand fetch.
+  // During the authentication, if jwks is not fetched or is expired, it will trigger
+  // an on-demand fetch. But async_fetch is preferred as it is done in the main thread
+  // and not need to block request processing but on-demand fetch is done in the worker
+  // thread and block request processing.
+  refetch_duration_ = getCacheDuration(remote_jwks);
+  if (refetch_duration_ > RefetchBeforeExpiredSec) {
+    refetch_duration_ = refetch_duration_ - RefetchBeforeExpiredSec;
```

Comment:
I am trying to understand the behavior here.

Let's say we configure the `remote_jwks.cache_duration` to 6 seconds then the `refetch_timer_` will be run for every for every 1 seconds (6 seconds - 5 seconds).

If we configure `remote_jwks.cache_duration` to 5 seconds then the `refetch_timer_` will be run for every 5 seconds. From documentation they will not be able to know this change in frequency.


Let's assume for practical purpose we would typically not set a cache duration of 10-5 seconds. But I am thinking how a user would make sense of this difference in behavior. Perhaps we should make that `RefetchBeforeExpiredSec` configurable, and let user know that if `cache_duration` is low (for Eg: 6 seconds), then also configure `RefetchBeforeExpiredSec` to something low (Eg: 1 second) so that fetch happens for every 5 seconds instead of 1 second

Reply:
Logically, you are correct,  there is a big jump between 1s retry for 6s cache_duration vs 4s retry for 4s cache_duration.  But I feel that normally cache_duration is around 10 minute or even hours,  rarely around 5 second range.    So spend too much time on such rare corner case is over kill.   We don't want to make thing too complicated with so many knots.  What do you think?

Reply:
understood